engine/src/asset/texture/mod.rs

```diff
@@ -0,0 +1,152 @@
+mod dat;
+
+use self::dat::texture::Texture;
+
+use super::{extension::*, AssetParser};
+use crate::utils::{compression::decompress, nom::*};
+use dat::{offset::TextureOffset, size::TextureSize, texture::MippedTexture};
+
+pub enum TextureMipKind {
+    NonMipped(Texture),
+    Mipped(MippedTexture),
+}
+
+pub enum TextureAnimationKind {
+    Static(TextureMipKind),
+    Animated(Vec<TextureMipKind>),
+}
+
+pub struct TextureOffsetCollection;
+
+impl AssetParser<Pack> for TextureOffsetCollection {
+    type Output = Vec<TextureOffset>;
+
+    type Context<'ctx> = ();
+
+    fn parser((): Self::Context<'_>) -> impl Fn(Input) -> Result<Self::Output> {
+        move |input| {
+            let (_, offsets) = multi::many0(TextureOffset::parser(()))(input)?;
+
+            Ok((&[], offsets))
+        }
+    }
+}
+
+pub struct MippedTextureCollection;
+
+impl AssetParser<Pack> for MippedTextureCollection {
+    type Output = Vec<TextureAnimationKind>;
+
+    type Context<'ctx> = &'ctx [TextureOffset];
+
+    fn parser(offsets: Self::Context<'_>) -> impl Fn(Input) -> Result<Self::Output> {
+        move |input| {
+            let textures = offsets
+                .iter()
+                .map(|o| {
+                    let input = &input[o.offset as usize..][..o.size_compressed as usize];
+                    let input = decompress(input);
+
+                    MippedTexture::parser(TextureSize {
+                        width: o.width,
+                        height: o.height,
+                    })(&input)
+                    .map(|(_, d)| (d, o))
+                })
+                .collect::<std::result::Result<Vec<_>, _>>()?;
+
+            let textures = textures
+                .iter()
+                .map(|(t, o)| {
+                    if o.animation_frames == 0 {
+                        TextureAnimationKind::Static(TextureMipKind::Mipped(t.clone()))
+                    } else {
+                        let mut frames = Vec::with_capacity(o.animation_frames as usize);
+                        let mut next = (t, *o);
+
+                        for i in 0..o.animation_frames {
+                            frames.push(TextureMipKind::Mipped(next.0.clone()));
+                            let (t, o) = &textures[next.1.next_animation_texture_id as usize];
+                            next = (t, o);
+                        }
+
+                        TextureAnimationKind::Animated(frames)
+                    }
+                })
+                .collect::<Vec<_>>();
```

Comment:
We can use `.cloned()` here, because we are cloning the texture inside `map`'s closure regardless. Also you can replace `mut next` variable with `fold`. If you think that it is confusing to shadow the outer `texture` and `offset` variables, you could rename the fold's parameters as `next_texture` and `next_offset`, respectively.

```suggestion
                                .cloned()
                .map(|(texture, offset)| {
                    if offset.animation_frames == 0 {
                        TextureAnimationKind::Static(TextureMipKind::Mipped(texture))
                    } else {
                        let mut frames = Vec::with_capacity(offset.animation_frames as usize);

                        (0..offset.animation_frames).fold(
                            (texture, offset),
                            |(texture, offset), _| {
                                frames.push(TextureMipKind::Mipped(texture));
                                textures[offset.next_animation_texture_id as usize].clone()
                            },
                        );

                        TextureAnimationKind::Animated(frames)
                    }
                })
```